File: java/leetcode/greedy/partition_labels.java
/**
 * 763. Partition Labels
 * A string S of lowercase letters is given. We want to partition this string into as many parts as possible so that each letter appears in at most one part, and return a list of integers representing the size of these parts.

Example 1:

Input: S = "ababcbacadefegdehijhklij"
Output: [9,7,8]
Explanation:
The partition is "ababcbaca", "defegde", "hijhklij".
This is a partition so that each letter appears in at most one part.
A partition like "ababcbacadefegde", "hijhklij" is incorrect, because it splits S into less parts.
Note:

S will have length in range [1, 500].
S will consist of lowercase letters ('a' to 'z') only.
 */

import java.util.List;
import java.util.ArrayList;


public class partition_labels {

    public List<Integer> partitionLabels(String S) {
        int[] lastIndexOfChar = new int[26];
        for (int i = 0; i < S.length(); i++) {
            lastIndexOfChar[char2Index(S.charAt(i))] = i;
        }
        List<Integer> partitions = new ArrayList<>();
        int firstIndex = 0;
        while (firstIndex < S.length()) {
            int lastIndex = firstIndex;
            for (int i = firstIndex; i < S.length() && i <= lastIndex; i++) {
                int index = lastIndexOfChar[char2Index(S.charAt(i))];
                if (index > lastIndex) {
                    lastIndex = index;
                }
            }
            partitions.add(lastIndex - firstIndex + 1);
            firstIndex = lastIndex+1;
        }
        return partitions;
    }

    private int char2Index(char c) {
        return c - 'a';
    }


}

File: golang/leetcode/strings/partition_labels.go
package strings

// 763. Partition Labels

func partitionLabels(S string) []int {
	lookup := make([]int, 26) // rune to index
	for i := 0; i < len(S); i++ {
		lookup[S[i]-'a'] = i
	}
	ret := make([]int, 0)
	firstIndex := 0
	for firstIndex < len(S) {
		lastIndex := firstIndex
		for i := firstIndex; i < len(S) && i <= lastIndex; i++ {
			index := lookup[S[i]-'a']
			if index > lastIndex {
				lastIndex = index
			}
		}
		ret = append(ret, lastIndex-firstIndex+1)
		firstIndex = lastIndex + 1
	}
	return ret
}
